splice_machine/src/main/java/com/splicemachine/derby/stream/iterator/merge/AbstractMergeJoinIterator.java

```diff
@@ -83,25 +88,42 @@ private int compare(ExecRow left, ExecRow right) throws StandardException {
         return 0;
     }
 
+    private int compareLefts(ExecRow lastLeft, ExecRow nextLeft) throws StandardException {
+        for (int i = 0, s = joinKeys.length; i < s; i = i + 2) {
+            int result = lastLeft.getColumn(joinKeys[i])
+                    .compare(nextLeft.getColumn(joinKeys[i]));
+            if (result != 0) {
+                return result;
+            }
+        }
+        return 0;
+    }
+
     protected Iterator<ExecRow> rightsForLeft(ExecRow left) throws IOException, StandardException {
         if (!currentRights.isEmpty() // Check to see if we've already collected the right rows
                 && compare(left, currentRights.get(0)) == 0){ // that match this left
             return currentRights.iterator();
         }
-        // If not, look for the ones that do
-        currentRights = new ArrayList<ExecRow>();
-        while (rightRS.hasNext()){
-            int comparison = compare(left, rightRS.peek());
-            if (comparison == 0) { // if matches left, add to buffer
-                currentRights.add(rightRS.next().getClone());
-            } else if (comparison < 0) { // if is greater than left, stop
-                break;
-            } else {
-                // if is less than left, read next right
-                rightRS.next();
+
+        if (leftRS.hasNext() && compareLefts(left,leftRS.peek()) == 0) { // Next Left
+            // If not, look for the ones that do
+            currentRights = new ArrayList<ExecRow>();
+            while (rightRS.hasNext()) {
+                int comparison = compare(left, rightRS.peek());
+                if (comparison == 0) { // if matches left, add to buffer
+                    currentRights.add(rightRS.next().getClone());
+                } else if (comparison < 0) { // if is greater than left, stop
+                    break;
+                } else {
+                    // if is less than left, read next right
+                    rightRS.next();
+                }
             }
```

Comment:
If left is the same as the next leftRow, we should have the right dataset already prepared. Why do we need to compute again?

Reply:
Please ignore this comment. My understanding was wrong. I think currentRights is prepared the second time the same left row key is seen.